app/models/search.rb

```diff
@@ -10,17 +10,29 @@ class Search < ApplicationRecord
   def self.where_terms(string)
     return all unless string
 
+    # Split each string into words by one-or-more whitespace characters
     sub_strings = string.strip.downcase.split(/\s+/).map do |str|
-      if str.start_with?("laa-")
-        clean_str = str.sub(/\Alaa-/, "laa")
-        "#{clean_str}:A"
-      elsif /\A\d+\/\d+\z/.match?(str) then "#{str}:A"
-      elsif /\A\d+{6}\z/.match?(str) then "#{str}:*"
+      # We escape with single quotes
+      str = str.gsub("'", "''")
+
+      # Is this an LAA reference or a UFN? (LAA-123ABC or 311223/001)
+      if str.start_with?("laa-") || /\A\d+\/\d+\z/.match?(str) then %('''#{str}''':A)
+
+      # Is this MAYBE a UFN? (6 digits for the first part, 9 digits for the full)
+      elsif /\A\d{6}\z/.match?(str) then "#{str}:*"
+      # Don't know what it is, but escape it and give it a lower weight anyway
       else
-        "#{str.tr('/', '-')}:*B"
+        %('''#{str}''':*B)
       end
     end
 
+    # The way the lexemes are setup in the search vector means a typical record ends up looking like
+    #
+    # '-123456':8A '010124/001':6A 'laa-123456':10A 'jason':2B 'jason/jim':1B 'jim':3B  'read':4B,5B
+    #
+    # In which we have the expected fields converted to lexemes, however in the case of the "Jason/Jim" first name; we generate lexemes for both the full and partial matches of both names.
+    #
+    # We only do this for defendant name as caseworkers use that syntax for alternate names.
```

Comment:
Nit: Are you saying that providers enter defendant names using "/" to indicate alternate names? If so, I can't see any evidence of that behaviour in our DB and don't remember any requirements around that. Or are you saying caseworkers _search_ for defendant names using "/"? In which case, shouldn't we have tests to verify that that works as expected in search?

Reply:
It was definitely raised at some point, but if it turns out that it's not actually a thing anymore that'll be very annoying as the bulk of the PR now is making that work...

And yeah there's already tests for alternate names, https://github.com/ministryofjustice/laa-crime-application-store/blob/main/spec/models/search_spec.rb#L178

Reply:
So those tests for alternate names specify that searching for "Jason/Jim" should only match defendants actually called "Jason/Jim" in the DB. In which case we shouldn't need to generate lexemes for "generate lexemes for both the full and partial matches of both names" should we? I think the goal of those tests was to demonstrate "the search doesn't break if there's a slash in the query", nothing more complex than that.

Reply:
Ah I think you're right, that's annoying.

I must have thankfully misremembered the full requirement, I'll remove the extra lexeme creation bits then.

Reply:
Wait one of those defendant names in the test is "Jason/Jim Read" @patrick-laa 

Does this mean the test is wrong or the requirement?

Reply:
We can't search for that though, we can only match either the whole token or the start of the token. I'm not sure why it was working _before_, but now that we have to escape the search query `/` is taken literally so while it doesn't break the query we can't use it without generating the other lexemes.

Reply:
Ah but first name and last name are separate tokens. This is why if the defendant is called "Jim Beam" you can search for "Beam" as well as searching for "Jim" - we shove first name, last name and firm name _separately_ into `search_fields`, so you can search on any one of them.

Reply:
You can, but when using `to_tsquery` raw as we were before, `-` is parsed as a word boundary so the `*` (match from start, _NOT_ partial matching) was able to work on both James and Jim.

Now we escape the query, special characters are no longer treated as word boundaries so all we can do is search the whole token or from the start of a token, making the "Jim" bit in "Jason/Jim" never resolve.

Reply:
But I have the test passing with your escaping but without any new extra lexeme code. The existing code replaces `/` with `-` in defendant names in the `search_fields` virtual column, and replaces `/` with `-` in the query. So `Json/Jim Read`in the DB becomes `json-jim:1B read:4B` in the virtual column, and `Jason/Jim` becomes `jason-jim` in the search query, and it just works. We don't need to add anything more to keep the tests passing.

Reply:
@elken aah, I finally understand the source of the confusion: That's because the `down` bit of your migration is buggy. If I apply your migration and then roll it back, I too end up with a bunch of failing tests. BUT I also end up with a schema that no longer resembles what it did before I first applied the migration. Whereas if instead I simply don't apply your migration in the first place and simply use your new code and new tests... they pass.

You've been led astray by a dodgy rollback!

Reply:
I'm afraid I'm not observing that either. With a fresh DB and no migration from this PR, all the LAA reference tests fail as the initial migration removes the `-` and now we don't do that.

Reply:
So look: All I'm doing is adding the escaping that you discovered, and all the new test cases you added, and all the James/Jim stuff still passes - https://github.com/ministryofjustice/laa-crime-application-store/pull/314/files

Reply:
@patrick-laa after our discussion and the latest changes are we happy with this now?

Reply:
So I feel like I should no longer be the arbiter of approval, but to me there's 3 things I'd consider changing:

1. The Ruby comment that this GH comment thread is attached to is inaccurate
2. You don't really need the clause on line 29-31 because you're still doing a tr on line 36 and we know that that tr already does everything that the lines 30-31 does, functionally speaking. 
3. I know that the "search for 'B' matches 'A/B' functionality isn't strictly needed, but since we already have that functionality and it already works, I don't think we really need to take the tests that exercise that functionality out.

Reply:
1. "caseworkers use that syntax for alternate names" - I'm not sure we have any evidence that they do
2. Without seeing the code from that CI run I can't comment on what went wrong for you, but look, here it is working without that extra clause: https://app.circleci.com/pipelines/github/ministryofjustice/laa-crime-application-store/1435/workflows/23f78080-c633-4c7c-b14a-382fed0e962b/jobs/9672
3. The test doesn't make an incorrect assumption about the system. It demonstrates that if the query is "Jim", and the defendant has been entered by the provider as having first name "Jason/Jim", searching for "Jim" wlll return the record. Which is true (hence why the test passes). It accurately describes the behaviour of the system. The fact that we don't need it to behave that way doesn't mean we want our tests to _less accurately_ describe how the system behaves.

Reply:
1. Oh you're right
2. I removed the lines you mentioned which are needed for the tests removed by point 3, but this is made moot if we just keep the tests around
3. I don't agree here, but keeping the tests renders point 2 moot so I'll just leave them